### PR TITLE
Fix `Path` and `Exception` content in the Problem Details Json

### LINKED ISSUE #43816: Problem details JSON returned by developer exception page serializes request path as an object
When enabling the new problem details service, the developer exception page middleware returns a JSON problem details to JSON-enabled clients but serializes the request path as an object instead of a string, likely due to the property on the payload inferring its type from `Request.Path` which is of type `PathString` (https://github.com/dotnet/aspnetcore/blob/main/src/Middleware/Diagnostics/src/DeveloperExceptionPage/DeveloperExceptionPageMiddlewareImpl.cs#L193)

image

Following up reproduces this when it receives a request to `/throw`:

```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddProblemDetails();

var app = builder.Build();

app.MapGet("/throw", () => throw new Exception("uh oh"));

app.Run();
```

@brunolins16

### PR DESCRIPTION
**Pull request by brunolins16**
Fixes https://github.com/dotnet/aspnetcore/issues/43816

## Description

`Developer Exception Middleware` generates a Problem Details with `Request Path` as `object`. This change update it to a `string`.

## Customer Impact

In Preview 7, the `Developer Exception Middleware` was updated to generate a Problem Details, all exception information including the `Request Path`.

``` json
"path": {
    "value": "/throw",
    "hasValue": true
}
```

With this change the `Problem Details` json payload will contains `Path` as `string` instead of `object`.

``` json
"path":  "/throw"
```

## Regression?

- [ ] Yes
- [x] No

It is actually a bug fix.

## Risk

- [ ] High
- [ ] Medium
- [x] Low

Very low risk, we're just fixing the generated Problem Details content introduced in .NET 7 Preview 7.

## Verification

- [X] Manual (required)
- [ ] Automated

## Packaging changes reviewed?

- [ ] Yes
- [ ] No
- [x] N/A

### CODE REVIEW

**Comment by BrennanConroy** [comment]
How do route values and headers look?

**Comment by brunolins16** [comment]
> How do route values and headers look?

Here is a full json:

```json
{
  "type": "https://tools.ietf.org/html/rfc7231#section-6.6.1",
  "title": "System.Exception",
  "status": 500,
  "detail": "Error message",
  "exception": {
    "details": [
      {
        "message": "Error message",
        "type": "System.Exception",
        "stackFrames": [
          {
            "function": "Program+<>c.<<Main>$>b__0_0(string name)",
            "file": "C:\\git\\ASPNETCore.Issues\\Issue43816\\Program.cs",
            "line": 21,
            "preContextLine": 15,
            "preContextCode": [
              "{",
              "    app.UseSwagger();",
              "    app.UseSwaggerUI();",
              "}",
              "",
              "app.MapGet(\"/throw/{name}\", (string name) => { "
            ],
            "contextCode": [
              "    throw new Exception(\"Error message\");"
            ],
            "postContextCode": [
              "});",
              "",
              "app.Run();"
            ],
            "errorDetails": null
          },
          {
            "function": "lambda_method2(Closure , object , HttpContext )",
            "file": null,
            "line": 0,
            "preContextLine": 0,
            "preContextCode": [],
            "contextCode": [],
            "postContextCode": [],
            "errorDetails": null
          },
          {
            "function": "Microsoft.AspNetCore.Routing.EndpointMiddleware.Invoke(HttpContext httpContext)",
            "file": null,
            "line": 0,
            "preContextLine": 0,
            "preContextCode": [],
            "contextCode": [],
            "postContextCode": [],
            "errorDetails": null
          },
          {
            "function": "Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)",
            "file": null,
            "line": 0,
            "preContextLine": 0,
            "preContextCode": [],
            "contextCode": [],
            "postContextCode": [],
            "errorDetails": null
          },
          {
            "function": "Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)",
            "file": null,
            "line": 0,
            "preContextLine": 0,
            "preContextCode": [],
            "contextCode": [],
            "postContextCode": [],
            "errorDetails": null
          },
          {
            "function": "Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context)",
            "file": "C:\\git\\aspnetcore\\src\\Middleware\\Diagnostics\\src\\DeveloperExceptionPage\\DeveloperExceptionPageMiddlewareImpl.cs",
            "line": 98,
            "preContextLine": 92,
            "preContextCode": [
              "    /// <param name=\"context\"></param>",
              "    /// <returns></returns>",
              "    public async Task Invoke(HttpContext context)",
              "    {",
              "        try",
              "        {"
            ],
            "contextCode": [
              "            await _next(context);"
            ],
            "postContextCode": [
              "        }",
              "        catch (Exception ex)",
              "        {",
              "            _logger.UnhandledException(ex);",
              "",
              "            if (context.Response.HasStarted)"
            ],
            "errorDetails": null
          }
        ]
      }
    ],
    "headers": {
      "Accept": [
        "*/*"
      ],
      "Connection": [
        "keep-alive"
      ],
      "Host": [
        "localhost:5266"
      ],
      "User-Agent": [
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36 Edg/105.0.1343.27"
      ],
      "Accept-Encoding": [
        "gzip, deflate, br"
      ],
      "Accept-Language": [
        "en-US,en;q=0.9"
      ],
      "Referer": [
        "http://localhost:5266/swagger/index.html"
      ],
      "sec-ch-ua": [
        "\"Microsoft Edge\";v=\"105\", \" Not;A Brand\";v=\"99\", \"Chromium\";v=\"105\""
      ],
      "sec-ch-ua-mobile": [
        "?0"
      ],
      "sec-ch-ua-platform": [
        "\"Windows\""
      ],
      "Sec-Fetch-Site": [
        "same-origin"
      ],
      "Sec-Fetch-Mode": [
        "cors"
      ],
      "Sec-Fetch-Dest": [
        "empty"
      ]
    },
    "error": "System.Exception: Error message\r\n   at Program.<>c.<<Main>$>b__0_0(String name) in C:\\git\\ASPNETCore.Issues\\Issue43816\\Program.cs:line 21\r\n   at lambda_method2(Closure, Object, HttpContext)\r\n   at Microsoft.AspNetCore.Routing.EndpointMiddleware.Invoke(HttpContext httpContext)\r\n--- End of stack trace from previous location ---\r\n   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)\r\n   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)\r\n   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context) in C:\\git\\aspnetcore\\src\\Middleware\\Diagnostics\\src\\DeveloperExceptionPage\\DeveloperExceptionPageMiddlewareImpl.cs:line 98",
    "path": "/throw/Bruno",
    "endpoint": "HTTP: GET /throw/{name}",
    "routeValues": {
      "name": "Bruno"
    }
  }
}
```

**Comment by BrennanConroy** [comment]
> How do route values and headers look?

Well those look fine, but that exception 😆

**Comment by halter73** [comment]
> > How do route values and headers look?
> 
> Well those look fine, but that exception 😆

Seriously though, we should update line 189 in this PR. The JSON serialized stack trace is beyond ridiculous.

```diff
                Details = exceptionDetails.Select(d => new
                {
                    Message = d.ErrorMessage ?? d.Error?.Message,
                    Type = TypeNameHelper.GetTypeDisplayName(d.Error),
-                    StackFrames = d.StackFrames,
+                    StackFrames = d.StackFrames.ToString(),
                }),
```

**Comment by brunolins16** [comment]
> > > How do route values and headers look?
> > 
> > 
> > Well those look fine, but that exception 😆
> 
> Seriously though, we should update line 189 in this PR. The JSON serialized stack trace is beyond ridiculous.
> 
> ```diff
>                 Details = exceptionDetails.Select(d => new
>                 {
>                     Message = d.ErrorMessage ?? d.Error?.Message,
>                     Type = TypeNameHelper.GetTypeDisplayName(d.Error),
> -                    StackFrames = d.StackFrames,
> +                    StackFrames = d.StackFrames.ToString(),
>                 }),
> ```

For this one we need to do more the ToString(). Let me see if there is something easy to do here.

```json
"details": [
      {
        "message": "Error message",
        "type": "System.Exception",
        "stackFrames": "System.Linq.Enumerable+SelectListIterator`2[Microsoft.Extensions.StackTrace.Sources.StackFrameInfo,Microsoft.Extensions.StackTrace.Sources.StackFrameSourceCodeInfo]"
      }
]
```

**Comment by brunolins16** [comment]
Which one is preferable?

```json
 "stackFrames": [
          {
            "location": "Program+<>c.<<Main>$>b__0_0(string name) in Program.cs, line 21",
            "code": [
              "15. {",
              "16.     app.UseSwagger();",
              "17.     app.UseSwaggerUI();",
              "18. }",
              "19. ",
              "20. app.MapGet(\"/throw/{name}\", (string name) => { ",
              "21.     throw new Exception(\"Error message\");",
              "22. });",
              "23. ",
              "24. app.Run();"
            ]
          }
]
```
or 

```json
"stackFrames": [
          "Program+<>c.<<Main>$>b__0_0(string name) in Program.cs, line 21\r\n15. {\r\n16.     app.UseSwagger();\r\n17.     app.UseSwaggerUI();\r\n18. }\r\n19. \r\n20. app.MapGet(\"/throw/{name}\", (string name) => { \r\n21.     throw new Exception(\"Error message\");\r\n22. });\r\n23. \r\n24. app.Run();\r\n",
          "lambda_method2(Closure , object , HttpContext )\r\n",
          "Microsoft.AspNetCore.Routing.EndpointMiddleware.Invoke(HttpContext httpContext)\r\n",
          "Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)\r\n",
          "Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)\r\n",
          "Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context) in DeveloperExceptionPageMiddlewareImpl.cs, line 98\r\n92.     /// <param name=\"context\"></param>\r\n93.     /// <returns></returns>\r\n94.     public async Task Invoke(HttpContext context)\r\n95.     {\r\n96.         try\r\n97.         {\r\n98.             await _next(context);\r\n99.         }\r\n100.         catch (Exception ex)\r\n101.         {\r\n102.             _logger.UnhandledException(ex);\r\n103. \r\n104.             if (context.Response.HasStarted)\r\n"
]
```

**Comment by halter73** [comment]
I don't like either of those formats. Mixing the `preContextCode`, `contextCode` and `postContextCode` into a `stackFrames` JSON array is really confusing. Especially when it's just smashed into the first string of an array.

I wonder if we shouldn't just do the following and call it a day:

```diff
-                Details = exceptionDetails.Select(d => new
-                {
-                    Message = d.ErrorMessage ?? d.Error?.Message,
-                    Type = TypeNameHelper.GetTypeDisplayName(d.Error),
-                    StackFrames = d.StackFrames,
-                }),
+                Details = new[] { errorContext.Exception.ToString() },
```

I don't see the value the ExceptionDetailsProvider is adding here. All the extra metadata about the stack frames is more noisy then helpful when all crammed into a JSON payload.

Is this what the output looked like before in .NET 6 when Developer Exception Middleware produced problem details for non-browser clients? If not, how does that look by comparison?

**Comment by BrennanConroy** [comment]
Error already contains the exception as a string, should be even be including the exception again in a potentially different format?

**Comment by brunolins16** [comment]
I don't think make sense format it, let me update to a simple `ToString()` and we are good to go 😁👏

**Comment by brunolins16** [comment]
Here is the updated payload:

```json
{
  "type": "https://tools.ietf.org/html/rfc7231#section-6.6.1",
  "title": "System.Exception",
  "status": 500,
  "detail": "Error message",
  "exception": {
    "details": "System.Exception: Error message\r\n   at Program.<>c.<<Main>$>b__0_0(String name) in C:\\git\\ASPNETCore.Issues\\Issue43816\\Program.cs:line 21\r\n   at lambda_method2(Closure, Object, HttpContext)\r\n   at Microsoft.AspNetCore.Routing.EndpointMiddleware.Invoke(HttpContext httpContext)\r\n--- End of stack trace from previous location ---\r\n   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)\r\n   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)\r\n   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddlewareImpl.Invoke(HttpContext context) in C:\\git\\aspnetcore\\src\\Middleware\\Diagnostics\\src\\DeveloperExceptionPage\\DeveloperExceptionPageMiddlewareImpl.cs:line 98",
    "headers": {
      "Accept": [
        "*/*"
      ],
      "Connection": [
        "keep-alive"
      ],
      "Host": [
        "localhost:5266"
      ],
      "User-Agent": [
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/105.0.0.0 Safari/537.36 Edg/105.0.1343.27"
      ],
      "Accept-Encoding": [
        "gzip, deflate, br"
      ],
      "Accept-Language": [
        "en-US,en;q=0.9"
      ],
      "Referer": [
        "http://localhost:5266/swagger/index.html"
      ],
      "sec-ch-ua": [
        "\"Microsoft Edge\";v=\"105\", \" Not;A Brand\";v=\"99\", \"Chromium\";v=\"105\""
      ],
      "sec-ch-ua-mobile": [
        "?0"
      ],
      "sec-ch-ua-platform": [
        "\"Windows\""
      ],
      "Sec-Fetch-Site": [
        "same-origin"
      ],
      "Sec-Fetch-Mode": [
        "cors"
      ],
      "Sec-Fetch-Dest": [
        "empty"
      ]
    },
    "path": "/throw/BRuno",
    "endpoint": "HTTP: GET /throw/{name}",
    "routeValues": {
      "name": "BRuno"
    }
  }
}
```

**Comment by brunolins16** [comment]
@dotnet/minimal-apis Can I have another review, please? 🙏

**Comment by DamianEdwards** [comment]
@brunolins16 if the header values are serialized as arrays (to support distinct multi-values) shouldn't the values there that have comma-separation appear as separate entries in the arrays? Perhaps it's better to just serialize the header values as strings.

**Comment by Tratcher** [comment]
> @brunolins16 if the header values are serialized as arrays (to support distinct multi-values) shouldn't the values there that have comma-separation appear as separate entries in the arrays? Perhaps it's better to just serialize the header values as strings.

@DamianEdwards it looks like the headers are being serialized the same way they were received from the client. I don't recommend trying to parse them into separate values if the client didn't send them that way. `User-Agent` shows why that would be non-trivial.

**Comment by brunolins16** [comment]
> @brunolins16 if the header values are serialized as arrays (to support distinct multi-values) shouldn't the values there that have comma-separation appear as separate entries in the arrays? Perhaps it's better to just serialize the header values as strings.

Not sure, because If i have multiple times the same header and it serializes correctly, eg.:

``` json
"Accept-Encoding": [
    "gzip, deflate, br",
    "custom"
]
```

**Comment by DamianEdwards** [comment]
So the array is there to support when headers are defined more than once? Seems like the wrong optimization given how uncommon that is.

**Comment by Tratcher** [comment]
It's not even a deliberate choice at the moment, that's just how we represent headers on the server. If we want to consolidate to a single string for this debug output that's fine.

**Review by halter73 — APPROVED** [review]

File: src/Middleware/Diagnostics/src/DeveloperExceptionPage/DeveloperExceptionPageMiddlewareImpl.cs [hunk 1]
```diff
@@ -179,18 +179,11 @@ private async Task DisplayExceptionContent(ErrorContext errorContext)
...
-            var exceptionDetails = _exceptionDetailsProvider.GetDetails(errorContext.Exception);
             problemDetails.Extensions["exception"] = new
             {
-                Details = exceptionDetails.Select(d => new
-                {
-                    Message = d.ErrorMessage ?? d.Error?.Message,
-                    Type = TypeNameHelper.GetTypeDisplayName(d.Error),
-                    StackFrames = d.StackFrames,
-                }),
+                Details = errorContext.Exception.ToString(),
```
Comment (halter73) [inline src/Middleware/Diagnostics/src/DeveloperExceptionPage/DeveloperExceptionPageMiddlewareImpl.cs:184]: There's nothing expecting details to be an array, right?

Comment (DamianEdwards) [reply src/Middleware/Diagnostics/src/DeveloperExceptionPage/DeveloperExceptionPageMiddlewareImpl.cs:184]: Don't believe so, this is simply developer-time formatting

**Review by BrennanConroy — APPROVED** [review]

**Comment by Pilchie** [comment]
Approved for .NET 7 RC2.

**Comment by brunolins16** [comment]
@dotnet/aspnet-build  Merge please. 🙏

**Comment by dougbu** [comment]
@brunolins16 please use `@aspnet-admins` to request servicing merges

**Pull request merged**